Refactor inbox to use materialized field instead of regexp

LucianBuzzo opened the pull request:
Previously the inbox had been disabled due to performance and stability
concerns. A chunk of this PR is simply re-enabling the inbox
functionality that as turned off in this PR https://github.com/product-os/jellyfish/pull/3568

Change-type: minor
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

***

**Please remember to write tests for your changes**. We aim to automatically
deploy `master` to production, and we can't safely do this without a solid
test suite!

LucianBuzzo committed:
Refactor inbox to use materialized field instead of regexp

Previously the inbox had been disabled due to performance and stability
concerns. A chunk of this PR is simply re-enabling the inbox
functionality that as turned off in this PR https://github.com/product-os/jellyfish/pull/3568

Change-type: minor
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

StefKors reviewed (commented):
Not sure if this should be fixed in this PR or in a new issue. 

Testing this PR I found out that when clicking the notification, it opens the channel but it doesn't modify the URL. More importantly, this makes it impossible to close the channel opened by a notification.

I'm taking a look at fixing this as it seems to be related to the generic `ADD_CHANNEL` store action. Hinting at that it would impact everything created by this action.

grahammcculloch committed:
ui: Ensure search works on all inbox tabs

Needed to be careful to avoid duplications when merging `required` fields - hence the use of `_.mergeWith` with a customizer function. Abstracting out the search filter to a single method will also prove useful when we start searching for `searchable` fields and the search filter has to be modified.

Change-type: patch
Signed-off-by: Graham McCulloch <graham@balena.io>

grahammcculloch reviewed (commented):
@LucianBuzzo I had a play with this branch locally and noticed a few issues to do with the search logic. I've pushed a commit that fixes these issues. If you're happy with that, ping me and I'll approve this PR.
